cclib/method/bader.py

```diff
@@ -0,0 +1,223 @@
+            )
+            self.chgdensity.data *= 2
+        else:
+            self.chgdensity = electrondensity_spin(
+                self.data,
+                self.volume,
+                [
+                    self.data.mocoeffs[0][: self.data.homos[0]],
+                    self.data.mocoeffs[1][: self.data.homos[1]],
+                ],
+            )
+
+        # Assign each grid point to Bader areas
+        self.fragresults = numpy.zeros(self.chgdensity.data.shape, "d")
+        nextIndex = 1
+
+        self.logger.info("Partitioning space into Bader areas.")
+
+        # Exclude the outermost positions
+        for xwalker in range(1, len(self.chgdensity.data) - 1):
+            for ywalker in range(1, len(self.chgdensity.data[0]) - 1):
+                for zwalker in range(1, len(self.chgdensity.data[0][0]) - 1):
+                    (xindex, yindex, zindex) = (xwalker, ywalker, zwalker)
+
+                    if self.fragresults[xindex, yindex, zindex] != 0:
+                        # index has already been assigned for this grid point
+                        continue
+                    else:
+                        listcoord = []
+                        loopcondition = True
+
+                        while loopcondition:
+                            # Here, `delta_rho` corresponds to equation 2,
+                            # and `grad_rho_dot_r` corresponds to equation 1 in the aforementioned
+                            # paper (doi:10.1016/j.commatsci.2005.04.010)
+                            delta_rho = (
+                                self.chgdensity.data[
+                                    xindex - 1 : xindex + 2,
+                                    yindex - 1 : yindex + 2,
+                                    zindex - 1 : zindex + 2,
+                                ]
+                                - self.chgdensity.data[xindex, yindex, zindex]
+                            )
+                            grad_rho_dot_r = delta_rho / _griddist
+                            maxat = numpy.where(grad_rho_dot_r == numpy.amax(grad_rho_dot_r))
+
+                            nextDirection = list(zip(maxat[0], maxat[1], maxat[2]))[0]
+                            nextDirection = [ind - 1 for ind in nextDirection]
+
+                            if len(list(zip(maxat[0], maxat[1], maxat[2]))) > 1:
+                                # when one or more directions indicate max grad (of 0), prioritize
+                                # to include all points in the Bader space
+                                if list(zip(maxat[0], maxat[1], maxat[2]))[0] == (1, 1, 1,):
+                                    nextDirection = list(zip(maxat[0], maxat[1], maxat[2]))[1]
+                                    nextDirection = [ind - 1 for ind in nextDirection]
+
+                            listcoord.append((xindex, yindex, zindex))
+
+                            if (
+                                self.fragresults[
```

Comment:
Can you store this outside the conditional as something like `bader_candidate_index` so you can reuse it inside the block?

Reply:
Yes -- I've just stored this and reused in the subsequent blocks. Thanks.